common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java

```diff
@@ -174,12 +191,47 @@ public static CalendarInterval fromDayTimeString(String s) throws IllegalArgumen
         int sign = m.group(1) != null && m.group(1).equals("-") ? -1 : 1;
         long days = m.group(2) == null ? 0 : toLongWithRange("day", m.group(3),
           0, Integer.MAX_VALUE);
-        long hours = toLongWithRange("hour", m.group(4), 0, 23);
-        long minutes = toLongWithRange("minute", m.group(5), 0, 59);
-        long seconds = toLongWithRange("second", m.group(6), 0, 59);
+        long hours;
+        long minutes;
+        long seconds;
+        if ("minute".equals(from)) {
+          hours = m.group(6) == null ? 0 : toLongWithRange("hour", m.group(4), 0, 23);
+          minutes = m.group(6) == null ? toLongWithRange("minute", m.group(4), 0, 59)
+            : toLongWithRange("minute", m.group(5), 0, 59);
+          seconds = m.group(6) == null ? toLongWithRange("minute", m.group(5), 0, 59)
+            : toLongWithRange("second", m.group(7), 0, 59);
+        } else {
+          if (m.group(8) != null && m.group(6) == null) {
+            hours = 0;
+            minutes = toLongWithRange("minute", m.group(4), 0, 59);
+            seconds = toLongWithRange("second", m.group(5), 0, 59);
+          } else {
+            hours = toLongWithRange("hour", m.group(4), 0, 23);
+            minutes = toLongWithRange("minute", m.group(5), 0, 59);
+            seconds = toLongWithRange("second", m.group(7), 0, 59);
+          }
+        }
         // Hive allow nanosecond precision interval
-        String nanoStr = m.group(8) == null ? null : (m.group(8) + "000000000").substring(0, 9);
+        String nanoStr = m.group(8) == null ? null : (m.group(9) + "000000000").substring(0, 9);
```

Comment:
? This line is wrong because `m.group(8) == null` and `m.group(9) + "000000000"`.
You need to be consistent.

Reply:
This is because the new dayTimePattern changed. The length of the pattern groups is 9.
 e.g. '20 15:40:32.99899999'  match(8) = ".99899999" and match(9) = "99899999"

Reply:
Oh. Got it. Thanks.

Reply:
Hmm. Wait. If `group(8) is null`, then `group(9) is null, too. You can use `group(9)` at `m.group(8) == null`, right?

Reply:
Yes, you are right.